scala-thoughts.md

```diff
@@ -0,0 +1,5 @@
+I find the balancd between object-oriented programming and functional programming is a strong point about Scala. It is fairly straightforward to create a new object in Scala just as it is in other object-oriented programming languages. Within definitions of objects, we can take advantage of the functional side of Scala. Based on my previous experience with functional programming languages, it was nice to be able to have and use higher-order functions. Pattern matching is also something I personally enjoyed using in Scala. Pattern matching is a strong point about Haskell, another functional programming language, that seems to be just as useful in Scala (although I don't know if there are any differences yet).
+
+Although Scala has loops like while and for loops, I feel that the language encourages the use of higher-order functions, recursion, and maybe partial-function application. For example, for loops tend to be iterations of collections of objects rather than using conditions and increments. It then feels like there are natural substitutions for any for loops placed into the language. In my code, I recognized that a for loop I was using to update a variable could have been replaced by a fold.
+
+I appreciate some of the features already included. The Options and the string methods provided such as slice are nice additions to the language. I would be curious if things in Scala analagous to functors, applicatives, and monoids in Haskell and whether we can leverage that to our use in the class.

```

Comment:
I wonder if Scala leaves it up to you to build these or if they are provided.  I'm very curious about what domains in which we might use these concepts from category theory.  Were there any particular use cases in any programming language in which you really wished you had applicatives, monoids, or monads available to solve a problem, or is this more of an academic consideration? Whether or not these category theoretical objects are provided or must be constructed.  An excellent question!